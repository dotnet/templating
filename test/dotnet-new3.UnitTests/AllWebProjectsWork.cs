// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using Microsoft.NET.TestFramework.Assertions;
using Microsoft.NET.TestFramework.Commands;
using Microsoft.TemplateEngine.TestHelper;
using Xunit;
using Xunit.Abstractions;

namespace Dotnet_new3.IntegrationTests
{
    public class AllWebProjectsWork : IClassFixture<AllProjectsWorkFixture>
    {
        private readonly AllProjectsWorkFixture _fixture;
        private readonly ITestOutputHelper _log;

        public AllWebProjectsWork(AllProjectsWorkFixture fixture, ITestOutputHelper log)
        {
            _fixture = fixture;
            _log = log;
        }

        [Theory]
        [InlineData("emptyweb_cs-50", "web")]
        [InlineData("mvc_cs-50", "mvc")]
        [InlineData("mvc_fs-50", "mvc", "-lang", "F#")]
        [InlineData("api_cs-50", "webapi")]
        [InlineData("emptyweb_cs-31", "web", "-f", "netcoreapp3.1")]
        [InlineData("mvc_cs-31", "mvc", "-f", "netcoreapp3.1")]
        [InlineData("mvc_fs-31", "mvc", "-lang", "F#", "-f", "netcoreapp3.1")]
        [InlineData("api_cs-31", "webapi", "-f", "netcoreapp3.1")]
        public void AllWebProjectsRestoreAndBuild(string testName, params string[] args)
        {
            string workingDir = Path.Combine(_fixture.BaseWorkingDirectory, testName);
            Directory.CreateDirectory(workingDir);

            new DotnetNewCommand(_log, args)
                .WithCustomHive(_fixture.HomeDirectory)
                .WithWorkingDirectory(workingDir)
                .Execute()
                .Should()
                .ExitWith(0)
                .And
                .NotHaveStdErr();

            new DotnetCommand(_log, "restore")
                .WithWorkingDirectory(workingDir)
                .Execute()
                .Should()
                .ExitWith(0)
                .And
                .NotHaveStdErr();

            new DotnetCommand(_log, "build")
                .WithWorkingDirectory(workingDir)
                .Execute()
                .Should()
                .ExitWith(0)
                .And
                .NotHaveStdErr();

            Directory.Delete(workingDir, true);
        }
    }

    public sealed class AllProjectsWorkFixture : SharedHomeDirectory
    {
        public AllProjectsWorkFixture(IMessageSink messageSink) : base(messageSink)
        {
            BaseWorkingDirectory = TestUtils.CreateTemporaryFolder(nameof(AllWebProjectsWork));
            // create nuget.config file with nuget.org listed
            new DotnetNewCommand(Log, "nugetconfig")
                .WithCustomHive(HomeDirectory)
                .WithWorkingDirectory(BaseWorkingDirectory)
                .Execute()
                .Should()
                .ExitWith(0)
                .And
                .NotHaveStdErr();

            InstallPackage("Microsoft.DotNet.Web.ProjectTemplates.5.0", BaseWorkingDirectory);
            InstallPackage("Microsoft.DotNet.Web.ProjectTemplates.3.1", BaseWorkingDirectory);
        }

        internal string BaseWorkingDirectory { get; private set; }
    }
}
