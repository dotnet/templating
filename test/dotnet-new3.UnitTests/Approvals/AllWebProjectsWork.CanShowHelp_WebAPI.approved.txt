ASP.NET Core Web API (C#)
Author: Microsoft
Description: A project template for creating an ASP.NET Core application with an example Controller for a RESTful HTTP service. This template can also be used for ASP.NET Core MVC Views and Controllers.
Options:                                                                                                                                                           
  -au|--auth                 The type of authentication to use                                                                                                     
                                 None             - No authentication                                                                                              
                                 IndividualB2C    - Individual authentication with Azure AD B2C                                                                    
                                 SingleOrg        - Organizational authentication for a single tenant                                                              
                                 Windows          - Windows authentication                                                                                         
                             Default: None                                                                                                                         

  --aad-b2c-instance         The Azure Active Directory B2C instance to connect to (use with IndividualB2C auth).                                                  
                             string - Optional                                                                                                                     
                             Default: https://login.microsoftonline.com/tfp/                                                                                       

  -ssp|--susi-policy-id      The sign-in and sign-up policy ID for this project (use with IndividualB2C auth).                                                     
                             string - Optional                                                                                                                     

  --aad-instance             The Azure Active Directory instance to connect to (use with SingleOrg auth).                                                          
                             string - Optional                                                                                                                     
                             Default: https://login.microsoftonline.com/                                                                                           

  --client-id                The Client ID for this project (use with SingleOrg or IndividualB2C auth).                                                            
                             string - Optional                                                                                                                     
                             Default: 11111111-1111-1111-11111111111111111                                                                                         

  --domain                   The domain for the directory tenant (use with SingleOrg or IndividualB2C auth).                                                       
                             string - Optional                                                                                                                     
                             Default: qualified.domain.name                                                                                                        

  --default-scope            The API scope the client needs to request to provision an access token. (use with IndividualB2C, SingleOrg).                          
                             string - Optional                                                                                                                     
                             Default: access_as_user                                                                                                               

  --tenant-id                The TenantId ID of the directory to connect to (use with SingleOrg auth).                                                             
                             string - Optional                                                                                                                     
                             Default: 22222222-2222-2222-2222-222222222222                                                                                         

  -r|--org-read-access       Whether or not to allow this application read access to the directory (only applies to SingleOrg auth).                               
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  --exclude-launch-settings  Whether to exclude launchSettings.json in the generated template.                                                                     
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  --no-https                 Whether to turn off HTTPS. This option only applies if Individual, IndividualB2C, SingleOrg, or MultiOrg aren't used for --auth.      
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  -uld|--use-local-db        Whether to use LocalDB instead of SQLite. This option only applies if --auth Individual or --auth IndividualB2C is specified.         
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  -f|--framework             The target framework for the project.                                                                                                 
                                 net5.0           - Target net5.0                                                                                                  
                                 netcoreapp3.1    - Target netcoreapp3.1                                                                                           
                                 netcoreapp2.1    - Target netcoreapp2.1                                                                                           
                             Default: net5.0                                                                                                                       

  --no-restore               If specified, skips the automatic restore of the project on create.                                                                   
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  --called-api-url           URL of the API to call from the web app. This option only applies if --auth SingleOrg or --auth IndividualB2C is specified.           
                             string - Optional                                                                                                                     
                             Default: https://graph.microsoft.com/v1.0                                                                                             

  --calls-graph              Specifies if the web app calls Microsoft Graph. This option only applies if --auth SingleOrg is specified.                            
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        

  --called-api-scopes        Scopes to request to call the API from the web app. This option only applies if --auth SingleOrg or --auth IndividualB2C is specified.
                             string - Optional                                                                                                                     
                             Default: user.read                                                                                                                    

  --no-openapi               Disable OpenAPI (Swagger) support                                                                                                     
                             bool - Optional                                                                                                                       
                             Default: false                                                                                                                        


To see help for other template languages (F#), use --language option:
   dotnet new3 webapi -h --language F#